File: ir/stmt/functiondef_ir.py
from dataclasses import dataclass

from common.span import SourceSpan
from common.typeparam_ir import TypeParamIR
from ir.annotation.annotation_ir import AnnotationIR
from ir.arg.arg_ir import ArgIR
from ir.expr.expr_ir import ExprIR
from generated import _pb2
from .stmt_ir import StmtIR, stmt_to_proto
from .decl_ir import DeclIR


@dataclass
class FunctionDefIR(DeclIR):
    id: int             # symbol id -> name of function
    symbol_id: int
    scope_id: int       # parent scope where function name is bound
    body_scope_id: int  # function-local scope
    name: str
    args: list[ArgIR]
    body: list[StmtIR]
    decorator_list: list[ExprIR]
    returns: AnnotationIR | None
    type_comment: str | None
    type_params: list[TypeParamIR]
    span: SourceSpan

    def to_proto(self):
        proto = _pb2.FunctionDefIR(
            id=self.id,
            symbol_id=self.symbol_id,
            name=self.name,
            scope_id=self.scope_id,
            body_scope_id=self.body_scope_id,
        )

        proto.args.extend([arg.to_proto() for arg in self.args])
        proto.body.extend([stmt_to_proto(stmt) for stmt in self.body])
        proto.decorator_list.extend([d.to_proto() for d in self.decorator_list])
        proto.type_params.extend([param.to_proto() for param in self.type_params])

        if self.returns is not None:
            proto.returns.CopyFrom(self.returns.to_proto())

        if self.type_comment is not None:
            proto.type_comment = self.type_comment

        if self.span is not None:
            proto.span.CopyFrom(self.span.to_proto())

        stmt = _pb2.DeclIR()
        stmt.function.CopyFrom(proto)
        return stmt

    def to_stmt_proto(self):
        decl = self.to_proto()
        stmt = _pb2.StmtIR()
        stmt.function.CopyFrom(decl.function)
        return stmt

File: rust/src/ir/stmt/functiondef_ir.rs
use crate::ir::nodes::TypeParamIR;
use crate::ir::{expr_ir::ExprIR, span_ir::SourceSpan, stmt_ir::StmtIR};
use crate::linker::symbol_ref::SymbolRef;

#[derive(Debug, Clone)]
pub enum ArgKind {
    PositionalOnly = 1,
    PositionalOrKeyword = 2,
    VarPositional = 3,
    KeywordOnly = 4,
    VarKeyword = 5,
}

impl TryFrom<i32> for ArgKind {
    type Error = String;

    fn try_from(int: i32) -> Result<Self, Self::Error> {
        match int {
            1 => Ok(ArgKind::PositionalOnly),
            2 => Ok(ArgKind::PositionalOrKeyword),
            3 => Ok(ArgKind::VarPositional),
            4 => Ok(ArgKind::KeywordOnly),
            5 => Ok(ArgKind::VarKeyword),
            _ => Err(format!("invalid ArgKind value: {}", int)),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ArgIR {
    pub symbol_id: usize,
    pub arg: String,
    pub kind: ArgKind,
    pub annotation: Option<ExprIR>,
    pub default: Option<Box<ExprIR>>,
    pub span: SourceSpan,
}

#[derive(Debug, Clone)]
pub struct ReturnIR {
    pub value: Option<Box<ExprIR>>,
    pub span: SourceSpan,
}

#[derive(Debug, Clone)]
pub struct FunctionDefIR {
    pub id: usize,
    pub symbol_id: usize,
    pub name: String,
    pub scope_id: usize,
    pub body_scope_id: usize,
    pub args: Vec<ArgIR>,
    pub body: Vec<StmtIR>,
    pub returns: Option<ExprIR>,
    pub decorator_list: Vec<ExprIR>,
    pub type_comment: Option<String>,
    pub type_params: Vec<TypeParamIR>,
    pub symbol_ref: SymbolRef,
    pub span: SourceSpan,
}
